File: Linq/RavenJArray.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Raven.Json.Utilities;

namespace Raven.Json.Linq
{
    public class RavenJArray : RavenJToken, IEnumerable<RavenJToken>
    {
        /// <summary>
		/// Initializes a new instance of the <see cref="RavenJArray"/> class.
        /// </summary>
        public RavenJArray()
        {
        }

        /// <summary>
		/// Initializes a new instance of the <see cref="RavenJArray"/> class from another <see cref="RavenJArray"/> object.
        /// </summary>
		/// <param name="other">A <see cref="RavenJArray"/> object to copy from.</param>
        public RavenJArray(RavenJArray other)
        {
            if (other.Length == 0) return;

            // clone array the hard way
            foreach (var item in other._items)
                Items.Add(item.CloneToken());
        }

		/// <summary>
		/// Initializes a new instance of the <see cref="RavenJArray"/> class with the specified content.
		/// </summary>
		/// <param name="content">The contents of the array.</param>
		public RavenJArray(params object[] content)
		{
			_items = new List<RavenJToken>();
			foreach (var item in content)
			{
				if (item.GetType().IsSubclassOf(typeof(RavenJToken)))
					_items.Add((RavenJToken)item);
				else
					_items.Add(new RavenJValue(item));
			}
		}

    	/// <summary>
        /// Gets the node type for this <see cref="RavenJToken"/>.
        /// </summary>
        /// <value>The type.</value>
        public override JTokenType Type
        {
            get { return JTokenType.Array; }
        }

        public override RavenJToken CloneToken()
        {
            return new RavenJArray(this);
        }

        public int Length { get { return _items == null ? 0 : _items.Count; } }

        public List<RavenJToken> Items
        {
            get { return _items ?? (_items = new List<RavenJToken>()); }
        }
        private List<RavenJToken> _items;

        internal new static RavenJArray Load(JsonReader reader)
        {
            if (reader.TokenType != JsonToken.StartArray)
                throw new Exception(
                    "Error reading JObject from JsonReader. Current JsonReader item is not an object: {0}".FormatWith(
                        CultureInfo.InvariantCulture, reader.TokenType));

            if (reader.Read() == false)
                throw new Exception("Unexpected end of json array");

            var ar = new RavenJArray();
            RavenJToken val = null;
            do
            {
                switch (reader.TokenType)
                {
                    case JsonToken.Comment:
                        // ignore comments
                        break;
                    case JsonToken.EndArray:
                        return ar;
                    case JsonToken.StartObject:
                        val = RavenJObject.Load(reader);
                        ar.Items.Add(val);
                        break;
                    case JsonToken.StartArray:
                        val = RavenJArray.Load(reader);
                        ar.Items.Add(val);
                        break;
                    default:
                        val = RavenJValue.Load(reader);
                        ar.Items.Add(val);
                        break;
                }
            } while (reader.Read());

            throw new Exception("Error reading JArray from JsonReader.");
        }

		/// <summary>
		/// Load a <see cref="RavenJArray"/> from a string that contains JSON.
		/// </summary>
		/// <param name="json">A <see cref="String"/> that contains JSON.</param>
		/// <returns>A <see cref="RavenJArray"/> populated from the string that contains JSON.</returns>
		public static new RavenJArray Parse(string json)
		{
			JsonReader jsonReader = new JsonTextReader(new StringReader(json));

			return Load(jsonReader);
		}

		/// <summary>
		/// Writes this token to a <see cref="JsonWriter"/>.
		/// </summary>
		/// <param name="writer">A <see cref="JsonWriter"/> into which this method will write.</param>
		/// <param name="converters">A collection of <see cref="JsonConverter"/> which will be used when writing the token.</param>
		public override void WriteTo(JsonWriter writer, params JsonConverter[] converters)
		{
			writer.WriteStartArray();

			if (_items != null)
			{
				foreach (var token in _items)
				{
					token.WriteTo(writer, converters);
				}
			}

			writer.WriteEndArray();
		}

		#region IEnumerable<RavenJToken> Members

		public IEnumerator<RavenJToken> GetEnumerator()
		{
			return Items.GetEnumerator();
		}

		#endregion

		#region IEnumerable Members

		System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		#endregion
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/json/linq/RavenJArray.java
package net.ravendb.abstractions.json.linq;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import net.ravendb.abstractions.exceptions.JsonReaderException;
import net.ravendb.abstractions.exceptions.JsonWriterException;

import org.codehaus.jackson.JsonFactory;
import org.codehaus.jackson.JsonGenerator;
import org.codehaus.jackson.JsonParser;
import org.codehaus.jackson.JsonToken;


/**
 * Represents a JSON array.
 */
public class RavenJArray extends RavenJToken implements Iterable<RavenJToken> {

  private boolean snapshot;

  private List<RavenJToken> items;

  /**
   * Initializes a new instance of the {@link RavenJArray} class.
   */
  public RavenJArray() {
    items = new ArrayList<>();
  }

  public RavenJArray(Collection<?> content) {
    items = new ArrayList<>();
    if (content == null) {
      return;
    }

    if (content instanceof RavenJToken && ((RavenJToken) content).getType().equals(JTokenType.ARRAY)) {

      items.add((RavenJToken) content);
    } else {
      for (Object item : content) {
        if (item instanceof RavenJToken) {
          items.add((RavenJToken) item);
        } else {
          items.add(new RavenJValue(item));
        }
      }
    }
  }

  /**
   * Initializes a new instance of the {@link RavenJArray} class with the specified content.
   * @param content The contents of the array;
   */
  public RavenJArray(RavenJToken... content) {
    this(Arrays.asList(content));
  }

  /**
   * Gets the node type for this {@link RavenJArray}
   */
  @Override
  public JTokenType getType() {
    return JTokenType.ARRAY;
  }

  /**
   * Gets the {@link RavenJToken} at the specified index.
   * @param index
   * @return
   */
  public RavenJToken get(int index) {
    return items.get(index);
  }

  /**
   * Sets the {@link RavenJToken} at the specified index.
   * @param index
   * @param value
   */
  public void set(int index, RavenJToken value) {
    checkSnapshot();
    items.set(index, value);
  }

  private void checkSnapshot() {
    if (snapshot) {
      throw new IllegalArgumentException("Cannot modify a snapshot, this is probably a bug.");
    }
  }

  @Override
  public RavenJArray cloneToken() {
    return (RavenJArray) cloneTokenImpl(new RavenJArray());
  }

  @Override
  public boolean isSnapshot() {
    return snapshot;
  }

  public int size() {
    return items.size();
  }

  public static RavenJArray load(JsonParser parser) {
    try {
      if (parser.getCurrentToken() == null) {
        if (parser.nextToken() == null) {
          throw new JsonReaderException("Error reading RavenJToken from JsonParser");
        }
      }
      if (parser.getCurrentToken() != JsonToken.START_ARRAY) {
        throw new JsonReaderException(
          "Error reading RavenJArray from JsonParser. Current JsonReader item is not an array: "
            + parser.getCurrentToken());
      }
      // advance to next token
      parser.nextToken();

      RavenJArray ar = new RavenJArray();
      RavenJToken val = null;
      do {
        switch (parser.getCurrentToken()) {
          case END_ARRAY:
            return ar;
          case START_OBJECT:
            val = RavenJObject.load(parser);
            ar.add(val);
            break;
          case START_ARRAY:
            val = RavenJArray.load(parser);
            ar.add(val);
            break;
          default:
            val = RavenJValue.load(parser);
            ar.add(val);
            break;
        }
      } while (parser.nextToken() != null);

      throw new JsonReaderException("Error reading RavenJArray from JsonReader.");

    } catch (IOException e) {
      throw new JsonReaderException(e.getMessage(), e);
    }
  }

  /**
   * Load a {@link RavenJArray} from a string that contains JSON.
   * @param json A {@link String} that contains JSON.
   * @return A {@link RavenJArray} populated from the string that contains JSON.
   */
  public static RavenJArray parse(String json) {
    try {
      JsonParser jsonParser = new JsonFactory().createJsonParser(json);
      return load(jsonParser);
    } catch (IOException e) {
      throw new JsonReaderException(e.getMessage(), e);
    }
  }

  /* (non-Javadoc)
   * @see java.lang.Iterable#iterator()
   */
  @Override
  public Iterator<RavenJToken> iterator() {
    return items.iterator();
  }

  public void add(RavenJToken token) {
    checkSnapshot();
    items.add(token);
  }

  public boolean remove(RavenJToken token) {
    checkSnapshot();
    return items.remove(token);
  }

  public void removeAt(int index) {
    checkSnapshot();
    items.remove(index);
  }

  /**
   * Inserts an item to the {@link List}   at the specified index.
   * @param index The zero-based index sat which item should be inserted
   * @param item The object to insert into the list.
   */
  public void insert(int index, RavenJToken item) {
    checkSnapshot();
    items.add(index, item);
  }

  /* (non-Javadoc)
   * @see raven.client.json.RavenJToken#ensureCannotBeChangeAndEnableShapshotting()
   */
  @Override
  public void ensureCannotBeChangeAndEnableShapshotting() {
    snapshot = true;
  }

  /* (non-Javadoc)
   * @see raven.client.json.RavenJToken#createSnapshot()
   */
  @Override
  public RavenJArray createSnapshot() {
    if (snapshot == false)
      throw new IllegalStateException("Cannot create snapshot without previously calling EnsureSnapShot");

    return new RavenJArray(items);
  }

  @Override
  public void writeTo(JsonGenerator writer) {
    try {
      writer.writeStartArray();

      if (items != null) {
        for (RavenJToken token : items) {
          token.writeTo(writer);
        }
      }

      writer.writeEndArray();
    } catch (IOException e) {
      throw new JsonWriterException(e.getMessage(), e);
    }
  }

  @Override
  protected void addForCloning(String key, RavenJToken token) {
    add(token);
  }

  @Override
  public Iterable<RavenJToken> values() {
    return items;
  }

  @Override
  public <T> List<T> values(Class<T> clazz) {
    return Extensions.convert(clazz, items);
  }

}
